extensions/2.0/Vendor/EXT_feature_metadata/README.md

```diff
@@ -0,0 +1,1691 @@
+
+Feature metadata may also be stored directly in textures. This is especially useful when texture mapping high frequency data, such as material properties, to less detailed 3D surfaces. Feature textures enable new styling and analytics capabilities, and complement glTF PBR textures.
+
+See [Examples](#examples) for a full list of use cases for this extension.
+
+## Feature Identification
+
+Features in a glTF primitive are identified in three ways:
+
+* Per-vertex using a vertex attribute
+* Per-texel using a glTF texture
+* Per-instance using an instance attribute with the [`EXT_mesh_gpu_instancing` extension](../../EXT_mesh_gpu_instancing)
+
+<img src="figures/metadata-access.png"  alt="Metadata Access" width="600">
+
+### Feature ID Vertex Attributes
+
+#### Feature ID Accessors
+
+The most straightforward method for defining feature IDs is to store them in a glTF vertex attribute. Feature ID attributes must follow the naming convention `_FEATURE_ID_X` where `X` is a non-negative integer. The first feature ID attribute is `_FEATURE_ID_0`, the second `_FEATURE_ID_1`, and so on.
+
+Feature IDs must be whole numbers in the range `[0, count - 1]` (inclusive), where `count` is the total number of features in the feature table.
+
+The attribute's accessor `type` must be `"SCALAR"` and `normalized` must be false. There is no restriction on `componentType`.
+
+> **Implementation Note:** since glTF accessors do not support `UNSIGNED_INT` types for 32-bit integers, `FLOAT` may be used instead. This allows for integer feature IDs up to 2²⁴. For smaller ranges of feature IDs, `UNSIGNED_BYTE` or `UNSIGNED_SHORT` can still be used. Note that this requires aligning each feature ID to 4-byte boundaries to adhere to glTF's alignment rules.
+
+![Feature Table](figures/feature-table.png)
+
+```jsonc
+{
+  "primitives": [
+    {
+      "attributes": {
+        "POSITION": 0,
+        "_FEATURE_ID_0": 1
+      },
+      "indices": 2,
+      "mode": 4,
+      "extensions": {
+        "EXT_feature_metadata": {
+          "featureIdAttributes": [
+            {
+              "featureTable": "buildings",
+              "featureIds": {
+                "attribute": "_FEATURE_ID_0"
+              }
+            }
+          ]
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Implicit Feature IDs
+
+In some cases it is possible to define feature IDs implicitly, such as when all vertices in a primitive have the same feature ID or when each vertex in a primitive has a different feature ID.
+
+This is accomplished by using `constant` and `divisor`.
+
+* `constant` sets a constant feature ID for each vertex. The default is `0`.
```

Comment:
> Were use cases beyond per-triangle values intended? Probably enough justification on its own but it'd be helpful to know.

Per-triangle features and per-point features are the main use cases here.

> could undefined divisor imply ∞ instead

Yeah, that makes sense to me

Reply:
Naming ideas, in no particular order:

- size
- featureSize
- featureLength
- repeat
- times

Reply:
Of these, I think "repeat" is the closest, as it's how many times each feature ID repeats. let's see:

* `constant: 0, repeat: 2` you get feature IDs: 0, 0, 1, 1, 2, 2, 3, 3, ...
* `constant: 1, repeat: 3` gives 1, 1, 1, 2, 2, 2, 3, 3, 3, ...
* `constant: 5, repeat: 0` gives 5, 5, 5, ...

feels pretty good to me

Reply:
>undefined divisor imply ∞ instead

@donmccurdy oh by this you mean the divisor is really infinity, not 0? since `Math.floor(i / Infinity)` technically works here, at least for floating point numbers

Reply:
and yeah you'd need to use something like `undefined`, as there's no infinity literal in JSON (nor do I think that's a good idea for understandability)

Reply:
I like `repeat` as well.

Is everyone good with `constant`? I have mixed feelings, when it's used with `repeat` it's no longer "constant".

Reply:
> oh by this you mean the divisor is really infinity, not 0?

I guess I'm thinking of it as a description of the resulting data, rather than as an input to the implied implementation. Implementations could also use the value in a loop invariant:

```js
// (a) using 'repeat' as divisor
const ids = new Float32Array(vertexCount).map((_, i) => Math.floor(i / repeat));

// (b) using 'repeat' as loop invariant
const ids = [];
for (let nextID = 0; ids.length < vertexCount; nextID++) {
  for (let j = 0; j < (repeat || vertexCount); j++) {
    ids.push(nextID);
  }
}
```

In words we might define this:

_"ID series begins at `constant`, and each ID value repeats `repeat` times before incrementing. If `repeat` is undefined, the ID does not increment and is repeated indefinitely."_

***

I have no objection to `constant`, the term `start` might be another option though.

Reply:
>"ID series begins at constant, and each ID value repeats repeat times before incrementing. If repeat is undefined, the ID does not increment and is repeated indefinitely."

I like this overall wording, though of course in the actual document use `feature ID` to be specific.

I feel a little on the fence, depending on how it's used I'd use one word over the other:

`start, repeat` -- they sound good together.
`constant, repeat`  -- doesn't exactly roll off the tongue but maybe not too bad given the way you explained it above
`constant` -- by itself (repeat undefined) makes sense
`start` -- by itself feels a bit weird, but I'd need to read the spec closely to understand that this means that every value is the same.

So feeling a bit iffy, but leaning slightly towards `constant`.

`offset` is another word, though conceptually I like `start` better if we're talking about the IDs rather than a byte offset.

Reply:
Resolving in https://github.com/CesiumGS/glTF/pull/13.